src/main/java/cs361/battleships/models/Ship.java

```diff
@@ -117,10 +114,21 @@ else if(attackedSquare.getIsCaptain()){
 
 		}
 	}
+
+	public boolean validateMove(char moveDir) {
+		int tRow = occupiedSquares.stream().max(Comparator.comparing(s -> s.getRow())).get().getRow();
```

Comment:
Looks like all of the bases are covered for this validation function. Good use of the stream on the occupiedSquares, simplifies the code a lot.